python/ray/experimental/dag/tests/test_dag_input.py

```diff
@@ -0,0 +1,252 @@
+"""
+Tests to ensure ray DAG can correctly mark its input(s) to take user
+request, for all DAGNode types.
+"""
+
+import pytest
+import ray.experimental.dag as ray_dag
+from typing import TypeVar
+
+import ray
+
+RayHandleLike = TypeVar("RayHandleLike")
+
+
+def test_simple_func(shared_ray_instance):
+    @ray.remote
+    def a(input: str):
+        return f"{input} -> a"
+
+    @ray.remote
+    def b(a: "RayHandleLike"):
+        # At runtime, a is replaced with execution result of a.
+        return f"{a} -> b"
+
+    # input -> a - > b -> ouput
+    a_node = a._bind(ray_dag.INPUT)
+    dag = b._bind(a_node)
+
+    assert ray.get(dag.execute("input")) == "input -> a -> b"
+    assert ray.get(dag.execute("test")) == "test -> a -> b"
+
+
+def test_func_dag(shared_ray_instance):
+    @ray.remote
+    def a(user_input):
+        return user_input
+
+    @ray.remote
+    def b(x):
+        return x * 2
+
+    @ray.remote
+    def c(x):
+        return x + 1
+
+    @ray.remote
+    def d(x, y):
+        return x + y
+
+    a_ref = a._bind(ray_dag.INPUT)
+    b_ref = b._bind(a_ref)
+    c_ref = c._bind(a_ref)
+    d_ref = d._bind(b_ref, c_ref)
+    d1_ref = d._bind(d_ref, d_ref)
+    d2_ref = d._bind(d1_ref, d_ref)
+    dag = d._bind(d2_ref, d_ref)
+    print(dag)
+
+    # [(2*2 + 2+1) + (2*2 + 2+1)] + [(2*2 + 2+1) + (2*2 + 2+1)]
+    assert ray.get(dag.execute(2)) == 28
+    # [(3*2 + 3+1) + (3*2 + 3+1)] + [(3*2 + 3+1) + (3*2 + 3+1)]
+    assert ray.get(dag.execute(3)) == 40
+
+
+def test_multi_input_func_dag(shared_ray_instance):
+    @ray.remote
+    def a(user_input):
+        return user_input * 2
+
+    @ray.remote
+    def b(user_input):
+        return user_input + 1
+
+    @ray.remote
+    def c(x, y):
+        return x + y
+
+    a_ref = a._bind(ray_dag.INPUT)
+    b_ref = b._bind(ray_dag.INPUT)
+    dag = c._bind(a_ref, b_ref)
+    print(dag)
+
+    # (2*2) + (2*1)
+    assert ray.get(dag.execute(2)) == 7
+    # (3*2) + (3*1)
+    assert ray.get(dag.execute(3)) == 10
+
+
+def test_invalid_input_as_class_constructor(shared_ray_instance):
+    @ray.remote
+    class Actor:
+        def __init__(self, val):
+            self.val = val
+
+        def get(self):
+            return self.val
+
+    with pytest.raises(
+        ValueError, match="INPUT cannot be used as ClassNode args"
+    ):
+        Actor._bind(ray_dag.INPUT)
```

Comment:
Actually, why can't it be used to bind here?

Reply:
Don't think this makes sense; we want to explicitly separate the "instantiate the DAG" and "execute the DAG on an input" steps (that's the purpose of having a static DAG definition based on my understanding)

Reply:
I guess it makes sense if we plan to add the separation in the future.

Reply:
I'm treating `ray_dag.INPUT` as user request sent to a pre-built, static DAG to execute, rather than a handy way to dynamically redefine a class constructor. We have a number of choices that can be considered "limiting" at first glance but probably better off guard-railed to begin with, such as enforcing JSON-serializable objects only in binding.